File: 学习与算法/openmp/1. openmp/2. constructs/lab_openmp/lab_openmp/daxpy.c
#include <stdio.h>
#include <sys/time.h>
#include <omp.h>
void suba(int N, double a[], double b[], double c[]);

int main(int argc, char *argv[])
{
   int N = 48*1024*1024;
   double a[N],b[N],c[N];
   
   suba(N,a,b,c);
}

void suba(int N, double a[], double b[], double c[])
{
   struct timeval it0,it1;
   double s,t,time,rate;
   int i,k;
   const int KMAX=10;

   for(i=0; i<N; ++i)
   {
      a[i] = 1.1;
      b[i] = 1.1;
   }
   s = 2.0;

   gettimeofday(&it0,NULL);

   for(k=0; k<KMAX; ++k)
   {
#pragma omp parallel for private(i,t), shared(a,b,c,N,s)
      for(i=0; i<N; ++i)
      {
         c[i] = a[i]*s+b[i];
      }
      t = t+c[N]+c[1];  //Keeps optimizer from removing outer loop
   }
   gettimeofday(&it1,NULL);
   long seconds = it1.tv_sec - it0.tv_sec;
   long microseconds = it1.tv_usec - it0.tv_usec;
   time = seconds + microseconds*1e-6;
   rate = KMAX*(double)(2*N)/((double)(1024*1024)*time);
   printf(" MFLOPS= %f    time= %f\n",rate,time);
   freopen("/dev/null","w",stdout);
   printf("%f\n",t);    //make sure t is used someplace
}

File: 学习与算法/openmp/1. openmp/2. constructs/lab_openmp/lab_openmp/daxpy.f90
program daxpy

   integer,parameter :: N=48*1024*1024
   real*8,save  :: a(N),b(N),c(N)

   call suba(N,a,b,c)

end program

subroutine suba(N,a,b,c)
   integer :: N
   real*8  :: a(N),b(N),c(N)
   real*8  :: s,t,time,rate
   integer :: irate,it0,it1,it2
   integer,parameter :: KMAX=10

   a=1.1;b=1.1;s=2.0

   call system_clock(count_rate=irate)
   call system_clock(it0)
   call system_clock(it1)

do k = 1,KMAX

!$OMP PARALLEL DO private(i,t), shared(a,b,c,N,s)
   do i = 1,N
      c(i) = a(i)*s+b(i)
   end do
   t = t+c(N)+c(1)  !Keeps optimizer from removing outer loop
end do

   call system_clock(it2)
   time = dble(it2 - it1*2 + it0)/dble(irate)
   rate = KMAX*dble(2*N)/(dble(1024*1024)*time)
   print*," MFLOPS=",rate,"    time=",time
   open(77,file='/dev/null')
   write(77,fmt='(d10.3)') t   ! make sure t is used someplace
   close(77)

end subroutine
